compiler/rustc_lint/src/non_local_def.rs

```diff
@@ -0,0 +1,171 @@
+use rustc_hir::{def::DefKind, Body, Item, ItemKind, Path, QPath, TyKind};
+use rustc_span::{def_id::DefId, sym, symbol::kw, MacroKind};
+
+use crate::{lints::NonLocalDefinitionsDiag, LateContext, LateLintPass, LintContext};
+
+declare_lint! {
+    /// The `non_local_definitions` lint checks for `impl` blocks and `#[macro_export]`
+    /// macro inside bodies (functions, enum discriminant, ...).
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// trait MyTrait {}
+    /// struct MyStruct;
+    ///
+    /// fn foo() {
+    ///     impl MyTrait for MyStruct {}
+    /// }
+    /// ```
+    ///
+    /// {{produces}}
+    ///
+    /// ### Explanation
+    ///
+    /// Creating non-local definitions go against expectation and can create decrepencies
+    /// in tooling. In should be avoided. It may become deny-by-default in edition 2024
+    /// and higher, see see the tracking issue <https://github.com/rust-lang/rust/issues/120363>.
+    ///
+    /// An an `impl` definition is non-local if it is nested inside an item and neither
+    /// the type nor the trait are at the same nesting level as the `impl` block.
+    ///
+    /// All nested bodies (functions, enum discriminant, array length, consts) (expect for
+    /// `const _: Ty = { ... }`, which is still undecided) are checked.
+    pub NON_LOCAL_DEFINITIONS,
+    Warn,
+    "checks for non-local definitions",
+    report_in_external_macro
+}
+
+#[derive(Default)]
+pub struct NonLocalDefinitions {
+    body_depth: u32,
+}
+
+impl_lint_pass!(NonLocalDefinitions => [NON_LOCAL_DEFINITIONS]);
+
+// FIXME(Urgau): Figure out how to handle modules nested in bodies.
+// It's currently not handled by the current logic because modules are not bodies.
+// They don't even follow the correct order (check_body -> check_mod -> check_body_post)
+// instead check_mod is called after every body has been handled.
+
+impl<'tcx> LateLintPass<'tcx> for NonLocalDefinitions {
+    fn check_body(&mut self, _cx: &LateContext<'tcx>, _body: &'tcx Body<'tcx>) {
+        self.body_depth += 1;
+    }
+
+    fn check_body_post(&mut self, _cx: &LateContext<'tcx>, _body: &'tcx Body<'tcx>) {
+        self.body_depth -= 1;
+    }
+
+    fn check_item(&mut self, cx: &LateContext<'tcx>, item: &'tcx Item<'tcx>) {
+        if self.body_depth == 0 {
+            return;
+        }
+
+        let parent_impl = cx.tcx.parent(item.owner_id.def_id.into());
+        let parent_impl_def_kind = cx.tcx.def_kind(parent_impl);
+
+        // Per RFC we (currently) ignore anon-const (`const _: Ty = ...`).
+        //
+        // FIXME: synstructure generate `const _DERIVE_krate_Trait_X_FOR_A`
+        // which is used by rustc so for now also ignore thoses.
+        if parent_impl_def_kind == DefKind::Const
```

Comment:
Annoyingly, I don't think this is enough. Test case:
```rust
trait Trait {}
struct T;
fn oh_no() {
    const _: () = { impl Trait for T {} }; //~ warn
}
```

We should probably recursively test that parents of the const are modules?.. Recursively to also prevent cases like
```rust
trait Trait {}
struct T;
fn oh_no() {
    mod hahahahahahahahaah {
        const _: () = { impl Trait for T {} }; //~ warn
    }
}
```

Reply:
okay, I pushed a fix for this; it doesn't handle nested `const _`, but it feels silly handle those